From 094daa07c441740a74d018ed478bd55faf9447c9 Mon Sep 17 00:00:00 2001
From: ktkachov <ktkachov@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Wed, 29 Oct 2014 21:56:01 +0000
Subject: [PATCH 326/326] [AArch64] Restore recog state after finding pre-madd
 instruction

    * config/aarch64/aarch64.c (aarch64_madd_needs_nop): Restore
    recog state after aarch64_prev_real_insn call.

    * gcc.target/aarch64/madd_after_asm_1.c: New test.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@216854 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog                                       |  5 +++++
 gcc/config/aarch64/aarch64.c                        |  4 ++++
 gcc/testsuite/ChangeLog                             |  4 ++++
 gcc/testsuite/gcc.target/aarch64/madd_after_asm_1.c | 14 ++++++++++++++
 4 files changed, 27 insertions(+)
 create mode 100644 gcc/testsuite/gcc.target/aarch64/madd_after_asm_1.c

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index ed79e0e..8d74c36 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,8 @@
+2014-10-29  Kyrylo Tkachov  <kyrylo.tkachov@arm.com>
+
+	* config/aarch64/aarch64.c (aarch64_madd_needs_nop): Restore
+	recog state after aarch64_prev_real_insn call.
+
 2014-10-24  Kyrylo Tkachov  <kyrylo.tkachov@arm.com>
 
 	* config.gcc (aarch64*-*-*): Define TARGET_FIX_ERR_A53_835769_DEFAULT
diff --git a/gcc/config/aarch64/aarch64.c b/gcc/config/aarch64/aarch64.c
index d756763..c262792 100644
--- a/gcc/config/aarch64/aarch64.c
+++ b/gcc/config/aarch64/aarch64.c
@@ -6148,6 +6148,10 @@ aarch64_madd_needs_nop (rtx insn)
     return false;
 
   prev = aarch64_prev_real_insn (insn);
+  /* aarch64_prev_real_insn can call recog_memoized on insns other than INSN.
+     Restore recog state to INSN to avoid state corruption.  */
+  extract_constrain_insn_cached (insn);
+
   if (!prev || !has_memory_op (prev))
     return false;
 
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index 195a3c8..65c98a7 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,7 @@
+2014-10-29  Kyrylo Tkachov  <kyrylo.tkachov@arm.com>
+
+	* gcc.target/aarch64/madd_after_asm_1.c: New test.
+
 2014-10-15  Eric Botcazou  <ebotcazou@adacore.com>
 
 	* gnat.dg/opt41.adb: New test.
diff --git a/gcc/testsuite/gcc.target/aarch64/madd_after_asm_1.c b/gcc/testsuite/gcc.target/aarch64/madd_after_asm_1.c
new file mode 100644
index 0000000..523941d
--- /dev/null
+++ b/gcc/testsuite/gcc.target/aarch64/madd_after_asm_1.c
@@ -0,0 +1,14 @@
+/* { dg-do assemble } */
+/* { dg-options "-O2 -mfix-cortex-a53-835769" } */
+
+int
+test (int a, double b, int c, int d, int e)
+{
+  double result;
+  __asm__ __volatile ("// %0, %1"
+                      : "=w" (result)
+                      : "0" (b)
+                      :    /* No clobbers */
+                      );
+  return c * d + e;
+}
-- 
2.0.3

